platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsParallelConfigurationIntegrationTest.groovy

```diff
@@ -82,5 +89,57 @@ class IsolatedProjectsParallelConfigurationIntegrationTest extends AbstractIsola
         outputDoesNotContain("Configure :b")
     }
 
+    def "task-graph listeners registered in parallel are all executed"() {
+        given:
+        def numberOfSubprojects = 10
+        def numberOfListenersPerProject = 10
+        def projects = (1..numberOfSubprojects).collect { "sub$it" }
+
+        projects.each {
+            settingsFile """
+                include("$it")
+            """
+
+            buildFile "$it/build.gradle", """
+                // call from background thread not to exhaust workers
+                new Thread() {
+                    @Override
+                    void run() {
+                        ${server.callFromBuildUsingExpression("'configure-' + project.name")}
```

Comment:
Doing it straight from the script body would mean that we had more projects (p) than workers (w), we would never see more than w requests, as there would be no workers left to run thew other scripts. The result would be a timeout waiting for the other (p - w) requests that we were expecting.

TBH, I don't think we would need to block on the HTTP server anyways.  As long as we can verify the `taskgraph.whenReady` listeners being invoked via the console output, we should be good. Right, @alllex ?

Reply:
It's a good catch about the number of workers, Rafael.

The test itself is not designed to prove that the implementation is thread-safe but to expose a case when an implementation is more obviously unsafe. So, it goes by ensuring multiple project evaluations (running in parallel) reach the same point in build logic and then start spamming the callback registration requests. The number of projects does not matter as much -- the important point is to get contention on callback registration. To that end, we still should have the blocking server, ensuring all projects are being configured in parallel and all of them are aligned before the "start line" and start spamming together.

Rather than blocking in an anonymous thread, I would resolve the issue with workers by making it match the number of projects (`--max-workers` ? and +1 for the root). It's okay if we want to reduce the number of projects for this, as we can, for instance, increase the number of listeners per project.

I would also leave a comment somewhere that this test is more of a soak test than a usual integration test, as it does not guarantee correctness, but instead detects some cases of unsafe implementations.

Reply:
Thanks for the clarification, Alex. Indeed, this test passing once does not guarantee correctness, but at the same time I think that is the case with many of our concurrency tests.

Reply:
I changed the test no to use ad hoc background threads